### PR TITLE
8273508: Support archived heap objects in SerialGC

### PR DESCRIPTION
**Pull request by iklam**
When `-XX:+UseSerialGC is enabled`, load the CDS archived heap objects into `SerialHeap::old_gen()` during VM bootstrap. This improves VM start-up time, mostly because the module graph can be loaded from the archive.

```
$ perf stat -r 40 java -XX:+UseSerialGC -version

Before: 0.042484507 seconds time elapsed ( +- 0.72% )
After:  0.031671000 seconds time elapsed ( +- 0.72% )
```

Changes in the gc subdirectories are contributed by @tschatzl

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8273508](https://bugs.openjdk.java.net/browse/JDK-8273508): Support archived heap objects in SerialGC

### Reviewers
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to 7d841aae314986e3fb0604bdaed926cda8a5d72a
 * [Calvin Cheung](https://openjdk.java.net/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to 7d841aae314986e3fb0604bdaed926cda8a5d72a

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5596/head:pull/5596` \
`$ git checkout pull/5596`

Update a local copy of the PR: \
`$ git checkout pull/5596` \
`$ git pull https://git.openjdk.java.net/jdk pull/5596/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5596`

View PR using the GUI difftool: \
`$ git pr show -t 5596`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5596.diff">https://git.openjdk.java.net/jdk/pull/5596.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/serial/tenuredGeneration.cpp [hunk 1]
```diff
@@ -219,6 +219,18 @@ void TenuredGeneration::object_iterate(ObjectClosure* blk) {
   _the_space->object_iterate(blk);
 }
 
+void TenuredGeneration::complete_loaded_archive_space(MemRegion archive_space) {
+  TenuredSpace* space = (TenuredSpace*)_the_space;
+
+  space->initialize_threshold();
+  HeapWord* start = archive_space.start();
+  while (start < archive_space.end()) {
+    size_t word_size = _the_space->block_size(start); /// Crashes here when accessing the klass
```
Comment (tschatzl) [inline src/hotspot/share/gc/serial/tenuredGeneration.cpp:228]: Please remove the debug comment :)
```suggestion
    size_t word_size = _the_space->block_size(start);
```

File: src/hotspot/share/gc/serial/tenuredGeneration.cpp [hunk 2]
```diff
@@ -219,6 +219,18 @@ void TenuredGeneration::object_iterate(ObjectClosure* blk) {
   _the_space->object_iterate(blk);
 }
 
+void TenuredGeneration::complete_loaded_archive_space(MemRegion archive_space) {
+  TenuredSpace* space = (TenuredSpace*)_the_space;
+
+  space->initialize_threshold();
```
Comment (tschatzl) [inline src/hotspot/share/gc/serial/tenuredGeneration.cpp:225]:
```suggestion
  // Create the BOT for the archive space.
  TenuredSpace* space = (TenuredSpace*)_the_space;
  space->initialize_threshold();
```

File: src/hotspot/share/gc/serial/serialHeap.cpp [hunk 3]
```diff
@@ -29,7 +29,9 @@
 #include "gc/shared/genMemoryPools.hpp"
 #include "gc/shared/strongRootsScope.hpp"
 #include "gc/shared/suspendibleThreadSet.hpp"
+#include "logging/log.hpp"
```
Comment (tschatzl) [inline src/hotspot/share/gc/serial/serialHeap.cpp:32]:
```suggestion
```
Debug code.

File: src/hotspot/share/gc/serial/serialHeap.cpp [hunk 4]
```diff
@@ -112,3 +114,14 @@ void SerialHeap::safepoint_synchronize_end() {
     SuspendibleThreadSet::desynchronize();
   }
 }
+
+HeapWord* SerialHeap::allocate_loaded_archive_space(size_t word_size) {
+  MutexLocker ml(Heap_lock);
+  HeapWord* result = old_gen()->allocate(word_size, /* is_tlab = */ false);
+  return result;
```
Comment (tschatzl) [inline src/hotspot/share/gc/serial/serialHeap.cpp:121]:
```suggestion
  return old_gen()->allocate(word_size, false /* is_tlab */);
```
Tighten the code...; GC code (see similar examples in the file) also adds the comment for bool parameters after that parameter.

**Review by tschatzl — COMMENTED** [review]
Initial pass over GC code.

**Review by tschatzl — APPROVED** [review]
GC changes (and the bits around them) seem good to me.

We already discussed the (existing) pervasive usage of `uintptr_t` in other code calling the gc code for addresses in the java heap and using size_t/int/intx/whatever for offsets in private which I do not recommend to do. However if there is any change to be made, it's a separate issue.

File: src/hotspot/share/cds/heapShared.cpp [hunk 5]
```diff
@@ -1739,9 +1757,11 @@ void HeapShared::verify_loaded_heap() {
 
 void HeapShared::fill_failed_loaded_region() {
   assert(_loading_failed, "must be");
-  HeapWord* bottom = (HeapWord*)_loaded_heap_bottom;
-  HeapWord* top = (HeapWord*)_loaded_heap_top;
-  Universe::heap()->fill_with_objects(bottom, top - bottom);
+  if (_loaded_heap_bottom != 0) {
+    HeapWord* bottom = (HeapWord*)_loaded_heap_bottom;
+    HeapWord* top = (HeapWord*)_loaded_heap_top;
```
Comment (calvinccheung) [inline src/hotspot/share/cds/heapShared.cpp:1762]: Maybe add `assert(_loaded_heap_top != 0, “must be”)` ?

**Review by calvinccheung — APPROVED** [review]
I've reviewed the CDS related changes. Just one minor comment below.

**Comment by iklam** [comment]
Thanks @tschatzl and @calvinccheung for the review.
/integrate

**Pull request closed without merging**